feat(cli/templates) add clojurescript

just-sultanov opened the pull request:
<!--
Update "[ ]" to "[x]" to check a box

Please make sure to read the Pull Request Guidelines: https://github.com/tauri-apps/tauri/blob/dev/.github/CONTRIBUTING.md#pull-request-guidelines
-->

### What kind of change does this PR introduce?
<!-- Check at least one. If you are introducing a new binding, you must reference an issue where this binding has been proposed, discussed and approved by the maintainers. -->

- [ ] Bugfix
- [x] Feature
- [ ] Docs
- [ ] New Binding issue #___
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

### Does this PR introduce a breaking change?
<!-- If yes, please describe the impact and migration path for existing applications in an attached issue. -->

- [ ] Yes, and the changes were approved in issue #___
- [x] No

### Checklist
- [x] When resolving issues, they are referenced in the PR's title (e.g `fix: remove a typo, closes #___, #___`)
- [x] A change file is added if any packages will require a version bump due to this PR per [the instructions in the readme](https://github.com/tauri-apps/create-tauri-app/blob/dev/.changes/readme.md).
- [x] I have added a convincing reason for adding this feature, if necessary

### Other information

Related to this nuxt/framework#184

just-sultanov committed:
feat(cli/templates): add clojurescript

amrbashir commented on packages/cli/fragments/fragment-clojurescript/_cta_manifest_:
```diff
@@ -0,0 +1,13 @@
+withGlobalTauri = true
+
+[files]
+cljs.svg = public/cljs.svg
```
remove this

amrbashir commented on packages/cli/fragments/_assets_/cljs.svg:
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0" encoding="utf-8"?>
```
move this file to `fragment-clojurescript/public`

amrbashir commented on packages/cli/fragments/fragment-clojurescript/shadow-cljs.edn:
```diff
@@ -0,0 +1,23 @@
+
+ :nrepl {:init-ns user}
+
+ :dev-http {3000 "public"
+            3001 "target/test"}
```
```suggestion
 :dev-http {1420"public"
            1421 "target/test"}
```

amrbashir commented on packages/cli/fragments/fragment-clojurescript/_cta_manifest_:
```diff
@@ -0,0 +1,13 @@
+
+beforeDevCommand = {{pkg_manager_run_command}} dev
+beforeBuildCommand = {{pkg_manager_run_command}} build
+devPath = http://localhost:3000
```
```suggestion
devPath = http://localhost:1420
```

amrbashir commented on README.md:
```diff
@@ -85,6 +85,7 @@ Currently supported template presets include:
 - `next-ts`
 - `preact`
 - `preact-ts`
+- `clojurescript`
```
you should do the same in `packages/cli/node/README.md`

amrbashir commented on .github/workflows/templates-test.yml:
```diff
@@ -115,6 +115,10 @@ jobs:
           sudo apt-get update
           sudo apt-get install -y libgtk-3-dev webkit2gtk-4.0 libayatana-appindicator3-dev librsvg2-dev patchelf
 
+      - run: |
+          sudo apt install -y clojure
+        if: matrix.settings.template == 'clojurescript'
+
```
move this up before `download cli` step

amrbashir commented on packages/cli/fragments/fragment-clojurescript/src/main/clojure/app/core.cljs:
```diff
@@ -0,0 +1,25 @@
+  []
+  (some->>
+   (gdom/getElement "root")
+   (dom/render [:h1 "Welcome to Tauri!"])))
```
This is a very very basic example, there should be an example showcasing how to invoke a rust command using https://tauri.app/v1/api/js/tauri#invoke

Also you need to match the same structure and functionality of the other templates which uses the css from `base/src/styles.css` and have a simple example for invoking rust command

amrbashir reviewed (changes requested):
You also need to add a post init note about installing `java` and `clojure`
A similar thing has been done for `yew` template in `packages/cli/src/templates.rs` in `post_init_info` method

Also don't forget to run `carg fmt --all` and `pnpm format` in the repo root

Thanks.

just-sultanov committed:
fix(cli/fragments/clojurescript): code review improvements

- add a post init note about installing `java` and `clojure`
- add an example using the `greet` command
- move some resources to the `clojurescript` template
- change ports for `shadow-cljs` and `tauri`
- update ci pipeline
- fix code formatting

amrbashir commented on packages/cli/fragments/fragment-clojurescript/src/test/clojure/app/core_test.cljs:
```diff
@@ -0,0 +1,9 @@
+(ns app.core-test
```
test related, remove

amrbashir commented on packages/cli/fragments/fragment-clojurescript/package.json:
```diff
@@ -0,0 +1,37 @@
+  "private": true,
+  "version": "0.0.0",
+  "scripts": {
+    "assets:build": "cp ./src/style.css ./public/",
```
why is this necessary? can't you import the style inside .cljs file directly?

just-sultanov replied:
Unfortunately, we cannot import styles into cljs

amrbashir replied:
then you need to move `base/src/style.css` to `_assets_/style.css` and add it to the `[files]` section in `_cta_manifest_` but you'd have to do the same for other fragments

just-sultanov replied:
Oh, super! Thanks 👍

amrbashir commented on packages/cli/fragments/fragment-clojurescript/package.json:
```diff
@@ -0,0 +1,37 @@
+    "shadow:watch": "npm run shadow-cljs watch app test",
+    "preshadow:build": "npm run assets:build",
+    "shadow:build": "npm run shadow-cljs release app",
+    "preshadow:test": "npm run assets:build",
+    "shadow:test": "npm run shadow-cljs compile ci && npm run karma start karma.config.js --single-run",
```
test related, remove this along with anything relates to testing in this file, depdencies or scripts

amrbashir commented on packages/cli/fragments/fragment-clojurescript/karma.config.js:
```diff
@@ -0,0 +1,22 @@
+      },
+    },
+    basePath: "target",
+    files: ["test/js/ci.js"],
+    frameworks: ["cljs-test"],
+    plugins: ["karma-cljs-test", "karma-chrome-launcher"],
```
test related, remove

amrbashir reviewed (changes requested):
I should've caught this in my review but the template shouldn't include testing so please remove all test related files/deps/scripts...etc

just-sultanov committed:
fix(cli/fragments/clojurescript): remove test stuff

just-sultanov committed:
refactor(cli/templates): move `base/src/style.css` to the `_assets_` directory

We can't import styles into cljs.

amrbashir commented on packages/cli/fragments/fragment-clojurescript/.vscode/extensions.json:
```diff
@@ -0,0 +1,3 @@
+{
+  "recommendations": ["tauri-apps.tauri-vscode", "rust-lang.rust-analyzer"]
```
Seems like clojurescript has a vscode extension, so we should recommend that
```suggestion
  "recommendations": ["tauri-apps.tauri-vscode", "rust-lang.rust-analyzer", "betterthantomorrow.calva"]
```

amrbashir commented on packages/cli/fragments/fragment-clojurescript/README.md:
```diff
@@ -0,0 +1,7 @@
+
+## Recommended IDE Setup
+
+- [VS Code](https://code.visualstudio.com/) + [Tauri](https://marketplace.visualstudio.com/items?itemName=tauri-apps.tauri-vscode) + [rust-analyzer](https://marketplace.visualstudio.com/items?itemName=rust-lang.rust-analyzer)
```
```suggestion
- [VS Code](https://code.visualstudio.com/) + [Tauri](https://marketplace.visualstudio.com/items?itemName=tauri-apps.tauri-vscode) + [rust-analyzer](https://marketplace.visualstudio.com/items?itemName=rust-lang.rust-analyzer) + [calva](https://marketplace.visualstudio.com/items?itemName=betterthantomorrow.calva)
```

amrbashir commented on packages/cli/fragments/fragment-clojurescript/deps.edn:
```diff
@@ -0,0 +1,9 @@
+ :deps {reagent/reagent {:mvn/version "1.1.1"
+                         :exclusions [cljsjs/react
+                                      cljsjs/react-dom
+                                      cljsjs/react-dom-server]}}
```
this package seems to be deprecated https://clojars.org/cljsjs/react-dom-server

just-sultanov replied:
This record means that we want to exclude some dependencies and use `React` directly from `npm`. But I double-checked and this record unnecessary. I've removed this part.

amrbashir commented on packages/cli/fragments/fragment-clojurescript/package.json:
```diff
@@ -0,0 +1,20 @@
+  "private": true,
+  "version": "0.0.0",
+  "scripts": {
+    "dev": "npm run shadow-cljs watch app",
+    "build": "npm run shadow-cljs release app",
```
```suggestion
    "dev": "shadow-cljs watch app",
    "build": "shadow-cljs release app",
```

amrbashir commented on packages/cli/fragments/fragment-clojurescript/src/main/clojure/app/core.cljs:
```diff
@@ -0,0 +1,54 @@
+    [reagent.core :as r]
+    [reagent.dom :as dom]))
+
+;;  Learn more about Tauri commands at https://tauri.app/v1/guides/features/command
```
Can you move this comment right above the `invoke` call?

amrbashir commented:
CI seems to be failing to build the clojure project

just-sultanov committed:
fix(cli/fragments/clojurescript): code review improvements

- added the recommended `calva` plugin for `vscode`
- updated deps
- removed exclusions for the `reagent` dependency

just-sultanov commented:
> CI seems to be failing to build the clojure project

I don't know why `deps.edn` can't be found. 
OK, I'll switch to using only `shadow-cljs` without `deps.edn`. Local testing of commands on `mac/ubuntu` passes without any problems.

just-sultanov committed:
refactor(cli/fragments/clojurescript): switch from `deps.edn` to `shadow-cljs` directly

just-sultanov commented:
@amrbashir all tests were passed. Only `pnpm` doesn't work.

Cause:
`The required JS dependency "object-assign" is not available, it has been requested "node_modules/react/cjs/react.production.min.js".`

I can fix it using the following methods:
1) Add  `.npmrc` to the root of the clojurescript template to allow install dependencies using `pnpm i --shamefully-hoist`
2) Add `object-assign` dependency to `package.json`
3) Support only `yarn` and `npm` package managers

What do you think about it?

nothingismagick commented:
In my opinion #2 seems the best, because #3 is the worst.

I read through similar troubles over at nuxt (maybe there is another solution?
https://github.com/nuxt/nuxt.js/issues/14146

amrbashir commented:
yeah option no.2 seems to be the best for now

just-sultanov committed:
fix(cli/fragments/clojurescript): fix issue with the `pnpm`

These settings allow us to use `pnpm` as a package manager.
Without these settings, we will receive several errors from the compiler:
`object-assign`, `scheduler`, `scheduler/tracing` is not available.

just-sultanov commented:
Unfortunately, option no.2 didn't work. I added a `.npmrc` file with the required parameters for the specified dependencies (without wildcards). In addition, I have added comments describing the problem with the compiler.

amrbashir commented:
That's fine, you can also create this file only if pnpm is the package manager so the file name would be like this `_[pnpm]_.npmrc`

just-sultanov commented:
> That's fine, you can also create this file only if pnpm is the package manager so the file name would be like this `_[pnpm]_.npmrc`

done

amrbashir commented:
Thank you

amrbashir commented on .changes/clojurescript.md:
```diff
@@ -0,0 +1,6 @@
+---
+"create-tauri-app": patch
+"create-tauri-app-js": patch
```
```suggestion
"create-tauri-app": minor
"create-tauri-app-js": minor
```

amrbashir committed:
Update .changes/clojurescript.md

phrohdoh commented:
It appears as if these changes were lost sometime between them being merged and the moment I'm writing this comment as the `.cljs` files, the `Template::ClojureScript` match arm, and the templated files are not on dev (bca3828ae1f10c476579b553213713d8e843908e).

Any idea what happened here?

FabianLars commented:
Many templates were removed in https://github.com/tauri-apps/create-tauri-app/pull/284
https://tauri.app/blog/2023/03/01/create-tauri-app-version-3-released#removing-templates
